doc/obxprop.1.in
.TH "OBXPROP" "1" 
.SH "NAME" 
openbox \(em A minimalistic, highly configurable, next generation window 
manager with extensive standards support. 
.SH "SYNOPSIS" 
.PP 
\fBopenbox\fR [\fB\-\-help\fP]  [\fB\-\-display\fP]  [DISPLAY]  [\fB\-\-id\fP]  [ID]  [\fB\-\-root\fP]  [\-\-]  [PROPERTY \&...]  
.SH "DESCRIPTION" 
.PP 
obxprop is a tool for displaying the properties on an x 
window. 
.PP 
This tool has a similar functionality to \fBxprop\fR, 
but obxprop allows you to see UTF-8 strings as text. 
.PP 
You may use the \fB\-\-id\fR option to specify a window 
identifier, otherwise obxprop will allow you to select a window by 
clicking on it. 
.PP 
Primarily, this tool exists for Openbox users to see the value of the 
_OB_NAME, _OB_CLASS, and _OB_ROLE properties, which Openbox uses for 
matching windows against user-defined application rules. 
.SH "OPTIONS" 
.PP 
These are the possible options that \fBopenbox\fR accepts: 
.IP "\fB\-\-help\fP" 10 
Show a summary of the options available. 
.IP "\fB\-\-display\fP" 10 
.IP "DISPLAY" 10 
Specify the X display to use. 
.IP "\fB\-\-id\fP" 10 
.IP "ID" 10 
Specify the window identifier for the window whose properties 
will be displayed. 
.PP 
Similar to \fBxprop\fR, 
you may pass one or more property names to have 
\fBopenbox\fR limit its output to only the properties 
requested. 
.SH "SEE ALSO" 
.PP 
openbox (1), openbox-session(1), openbox-gnome-session(1), 
openbox-kde-session(1). 
.PP 
Please report bugs to: \fBhttp://bugzilla.icculus.org/ 
\fP 
.\" created by instant / docbook-to-man, Mon 08 Feb 2010, 14:21 


CHANGELOG
3.4.11:
  * Update Hungarian, Japanese, and Latvian translations.
  * Make xdg-autostart use the OPENBOX environment by default, so you can use
    OnlyShowIn=OPENBOX in an autostart .desktop and it will work as expected.
  * Don't close the menu when you hold control and execute something.
  * Fix bug #4503 (Adjust who shows up in the Alt-Tab list using SKIP_TASKBAR).
  * Fix flickering window when moving maximized window between monitors of
    different sizes.
  * Fix bug #4355 (Allow multiple escaped _'s in a menu label and allow
    a menu shortcut to come after an escaped _).
  * Remember the maximized state of a window when it goes fullscreen, and
    restore it when leaving fullscreen state.
  * Fix bug #4072 (Openbox is stopped by terminal applications writing to
    stdout).
  * Fix bug #4492 (Mistake in openbox-gnome-session check while setting up).
  * Fix obxprop to make --root and --id work correctly.
  * Add _OB_APP_ROLE/CLASS/NAME/TYPE properties (replaces _OB_ROLE/CLASS/NAME).
  * Make the focus cycling popup dynamic when windows appear/disappear.
  * Fix bug #4411 (Crash when window appears during focus cycling).
  * Allow the user to specify which properties should be shown by obxprop.
  * Fix tilde expansion in the Execute action
  * Make Home and End keys move to the top/bottom of the active menu.
  * Use the submenuShowDelay when navigating menus with the keyboard.

3.4.10:
  * Improve keyboard navigation in Openbox menus.
  * Add a --root option and a manpage for obxprop.
  * Use a negative value for submenuShowDelay and submenuHideDelay to cause
    an infinite delay.  This means you have to click to show a submenu, rather
    than just hover over it.
  * Improved code for submenu show/hide delay.  Added the submenuHideDelay
    config file option, under the "menu" section.
  * Fixed bug #4464 (Typo in openbox-gnome-session script).
  * Fixed bug #4436 (Focusing a window used to stop focus cycling).
  * Renamed obprop to obxprop due to collision with Open Babel (See bug #4419).

3.4.9:
  * Allow focus to move while inside an Openbox menu, or during an interactive
    action such as window cycling.
  * Fixed bug #3717 (Empty dock interfered with move/grow to edge actions).
  * Fixed bug #4411 (Crash when switching desktops and window cycling).
  * Fixed bug #4377 (Window resistance against struts).
  * Fixed bug #4035 (Prevent focus from moving under the mouse after
    activating a window with an Openbox menu.
  * Correct the value provided by the _NET_WORKAREA hint, so desktop icons
    will place across all monitors.
  * Don't hide submenus immediately when moving through the parent menu.
    (Resolves request #3762).
  * Fix for showing Openbox menus with multiple monitors, don't restrict them
    to the monitor where the mouse is.
  * Fixed bug #4023 (Allow the user to have multiple keys which perform the
    same function in Openbox menus/move/resize.  E.g. two keys which are
    both bound to Escape.
  * Add a new obprop tool, which can be used to read the _OB_ROLE, _OB_NAME,
    and _OB_CLASS (as well as any other UTF-8 window properties) off of a
    window.
  * Add _OB_ROLE, _OB_NAME, and _OB_CLASS hints on each window that show the
    respective values for use in the rc.xml applications section, to modify
    the window when it appears.
  * Improve Openbox interoperability with gnome-session >= 2.24.
  * Fixed bug #4040 (Remove desktop hints set by gdm in the openbox-session
    scripts, so that Openbox can set the number of desktops (assuming
    gnome-settings-daemon doesn't first)).
  * Fix a bug in xdg-autostart preventing some .desktop files from working.
  * Show the desktop pager popup on the primary monitor instead of on all
    monitors.
  * Add a new primaryMonitor config option, which is where Openbox popups
    will appear.  Defaults to a fixed monitor, but can be made to behave as
    before with the "active" value for it.
  * Correct edge detection for move/grow to edge to properly use monitor edges
    for multi-monitor setups.
  * Change default window placement policy to stay on the active monitor for
    multi-monitor setups.
  * Fixed bug #1149 (Crash with some window icon sizes).
  * Respond to all strut changes, fixes moving/hiding panels.
  * Fix internal code to focus windows on other desktops correctly (Fixes
    bug #4413).
  * Focus correctness fixups for switching desktops.
  * Fixed bug #4373 (Decoration bug for shaded maximized windows).
  * Fixed bug #4350 (Allow a window to be made skip_taskbar but still get
    focused by the user's rc.xml).
  * Fixed bug #4307 (Set a minimum time for screenEdgeWarpTime).
  * Fixed bug #4253 (Support for Solaris in openbox-session scripts).
  * Fixed bug #3851 (Allow transient windows to be above helper windows).

3.4.8:
  * Updated translations: Slovak.
  * Allow windows to change their decorations at any time (Fix for
    Google Chrome).
  * Make openbox-session to respect the $XDG_CONFIG_HOME environment variable.
  * Fixed bug #4344 where borders were given to windows that should not have
    them.
  * Merge the SessionLogout and Exit actions.  They now test if connected to a
    session manager and ask it to exit if so, or simply kill Openbox if not.
  * Further tweaks to the _NET_ACTIVE_WINDOW message handling.  Use the same
    logic for focus-stealing as is used when mapping a new window.
  * Don't go out of our way any more to prevent focus from moving while the
    keyboard is grabbed.
  * Fix openbox-gnome-session when using gnome-session > 2.22.

3.4.8-rc2:
  * Updated translations: Italian, Croatian, Ukrainian.
  * When resizing a window while focus cycling with bar=no, the bar no longer
    reappears.
  * Correctly handle shaped windows using the ShapeInput kind, this is used
    by many composited apps to pass through clicks in their transparent areas.
  * Fix the <monitor> per-app setting.
  * Avoid using anonymous unions.

tools/obxprop/obxprop.c
#include <X11/Xlib.h>
#include <X11/cursorfont.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <assert.h>
#include <glib.h>

gint fail(const gchar *s) {
    if (s)
        fprintf(stderr, "%s\n", s);
    else
        fprintf
            (stderr,
             "Usage: obxprop [OPTIONS] [--] [PROPERTIES ...]\n\n"
             "Options:\n"
             "    --help              Display this help and exit\n"
             "    --display DISPLAY   Connect to this X display\n"
             "    --id ID             Show the properties for this window\n"
             "    --root              Show the properties for the root window\n");
    return 1;
}

gint parse_hex(gchar *s) {
    gint result = 0;
    while (*s) {
        gint add;
        if (*s >= '0' && *s <='9')
            add = *s-'0';
        else if (*s >= 'A' && *s <='F')
            add = *s-'A';
        else if (*s >= 'a' && *s <='f')
            add = *s-'a';
        else
            break;

        result *= 16;
        result += add;
    }
    return result;
}

Window find_client(Display *d, Window win)
{
    Window r, *children;
    guint n, i;
    Atom state = XInternAtom(d, "WM_STATE", True);
    Atom ret_type;
    gint ret_format, res;
    gulong ret_items, ret_bytesleft, *xdata;

    XQueryTree(d, win, &r, &r, &children, &n);
    for (i = 0; i < n; ++i) {
        Window w = find_client(d, children[i]);
        if (w) return w;
    }

    // try me
    res = XGetWindowProperty(d, win, state, 0, 1,
                       False, state, &ret_type, &ret_format,
                       &ret_items, &ret_bytesleft,
                       (unsigned char**) &xdata);
    XFree(xdata);
    if (res != Success || ret_type == None || ret_items < 1)
        return None;
    return win; // found it!
}

static gboolean get_all(Display *d, Window win, Atom prop,
                        Atom *type, gint *size,
                        guchar **data, guint *num)
{
    gboolean ret = FALSE;
    gint res;
    guchar *xdata = NULL;
    gulong ret_items, bytes_left;

    res = XGetWindowProperty(d, win, prop, 0l, G_MAXLONG,
                             FALSE, AnyPropertyType, type, size,
                             &ret_items, &bytes_left, &xdata);
    if (res == Success) {
        if (ret_items > 0) {
            guint i;

            *data = g_malloc(ret_items * (*size / 8));
            for (i = 0; i < ret_items; ++i)
                switch (*size) {
                case 8:
                    (*data)[i] = xdata[i];
                    break;
                case 16:
                    ((guint16*)*data)[i] = ((gushort*)xdata)[i];
                    break;
                case 32:
                    ((guint32*)*data)[i] = ((gulong*)xdata)[i];
                    break;
                default:
                    g_assert_not_reached(); /* unhandled size */
                }
            *num = ret_items;
            ret = TRUE;
        }
        XFree(xdata);
    }
    return ret;
}

gchar *append_string(gchar *before, gchar *after, gboolean quote)
{
    gchar *tmp;
    const gchar *q = quote ? "\"" : "";
    if (before)
        tmp = g_strdup_printf("%s, %s%s%s", before, q, after, q);
    else
        tmp = g_strdup_printf("%s%s%s", q, after, q);
    g_free(before);
    return tmp;
}

gchar *append_int(gchar *before, guint after)
{
    gchar *tmp;
    if (before)
        tmp = g_strdup_printf("%s, %u", before, after);
    else
        tmp = g_strdup_printf("%u", after);
    g_free(before);
    return tmp;
}

gchar* read_strings(gchar *val, guint n, gboolean utf8)
{
    GSList *strs = NULL, *it;
    gchar *ret, *p;
    guint i;

    p = val;
    while (p < val + n) {
        strs = g_slist_append(strs, g_strndup(p, n - (p - val)));
        p += strlen(p) + 1; /* next string */
    }

    ret = NULL;
    for (i = 0, it = strs; it; ++i, it = g_slist_next(it)) {
        char *data;

        if (utf8) {
            if (g_utf8_validate(it->data, -1, NULL))
                data = g_strdup(it->data);
            else
                data = g_strdup("");
        }
        else
            data = g_locale_to_utf8(it->data, -1, NULL, NULL, NULL);

        ret = append_string(ret, data, TRUE);
        g_free(data);
    }

    while (strs) {
        g_free(strs->data);
        strs = g_slist_delete_link(strs, strs);
    }
    return ret;
}

gchar* read_atoms(Display *d, guchar *val, guint n)
{
    gchar *ret;
    guint i;

    ret = NULL;
    for (i = 0; i < n; ++i)
        ret = append_string(ret, XGetAtomName(d, ((guint32*)val)[i]), FALSE);
    return ret;
}

gchar* read_numbers(guchar *val, guint n, guint size)
{
    gchar *ret;
    guint i;

    ret = NULL;
    for (i = 0; i < n; ++i)
        switch (size) {
        case 8:
            ret = append_int(ret, ((guint8*)val)[i]);
            break;
        case 16:
            ret = append_int(ret, ((guint16*)val)[i]);
            break;
        case 32:
            ret = append_int(ret, ((guint32*)val)[i]);
            break;
        default:
            g_assert_not_reached(); /* unhandled size */
        }

    return ret;
}

gboolean read_prop(Display *d, Window w, Atom prop, const gchar **type, gchar **val)
{
    guchar *ret;
    guint nret;
    gint size;
    Atom ret_type;

    ret = NULL;
    if (get_all(d, w, prop, &ret_type, &size, &ret, &nret)) {
        *type = XGetAtomName(d, ret_type);

        if (strcmp(*type, "STRING") == 0)
            *val = read_strings((gchar*)ret, nret, FALSE);
        else if (strcmp(*type, "UTF8_STRING") == 0)
            *val = read_strings((gchar*)ret, nret, TRUE);
        else if (strcmp(*type, "ATOM") == 0) {
            g_assert(size == 32);
            *val = read_atoms(d, ret, nret);
        }
       else
            *val = read_numbers(ret, nret, size);

        g_free(ret);
        return TRUE;
    }
    return FALSE;
}

void show_properties(Display *d, Window w, int argc, char **argv)
{
    Atom* props;
    int i, n;

    props = XListProperties(d, w, &n);

    for (i = 0; i < n; ++i) {
        const char *type;
        char *name, *val;

        name = XGetAtomName(d, props[i]);

        if (read_prop(d, w, props[i], &type, &val)) {
            int found = 1;
            if (argc) {
                int i;

                found = 0;
                for (i = 0; i < argc; i++)
                    if (!strcmp(name, argv[i])) {
                        found = 1;
                        break;
                    }
            }
            if (found)
                g_print("%s(%s) = %s\n", name, type, val);
            g_free(val);
        }

        XFree(name);
    }

    XFree(props);
}

int main(int argc, char **argv)
{
    Display *d;
    Window id, userid = None;
    int i;
    char *dname = NULL;
    gboolean root = FALSE;

    for (i = 1; i < argc; ++i) {
        if (!strcmp(argv[i], "--help")) {
            return fail(NULL);
        }
        else if (!strcmp(argv[i], "--root"))
            root = TRUE;
        else if (!strcmp(argv[i], "--id")) {
            if (++i == argc)
                return fail(NULL);
            if (argv[i][0] == '0' && argv[i][1] == 'x') {
                /* hex */
                userid = parse_hex(argv[i]+2);
            }
            else {
                /* decimal */
                userid = atoi(argv[i]);
            }
            if (!userid)
                return fail("Unable to parse argument to --id.");
        }
        else if (!strcmp(argv[i], "--display")) {
            if (++i == argc)
                return fail(NULL);
            dname = argv[i];
        }
        else if (*argv[i] != '-')
            break;
        else if (!strcmp(argv[i], "--")) {
            i++;
            break;
        }
        else
            return fail(NULL);
    }

    d = XOpenDisplay(dname);
    if (!d) {
        return fail("Unable to find an X display. "
                    "Ensure you have permission to connect to the display.");
    }

    if (root)
        userid = RootWindow(d, DefaultScreen(d));

    if (userid == None) {
        int j;
        j = XGrabPointer(d, RootWindow(d, DefaultScreen(d)),
                         False, ButtonPressMask,
                         GrabModeAsync, GrabModeAsync,
                         None, XCreateFontCursor(d, XC_crosshair),
                         CurrentTime);
        if (j != GrabSuccess)
            return fail("Unable to grab the pointer device");
        while (1) {
            XEvent ev;

            XNextEvent(d, &ev);
            if (ev.type == ButtonPress) {
                XUngrabPointer(d, CurrentTime);
                userid = ev.xbutton.subwindow;
                break;
            }
        }
        id = find_client(d, userid);
    }
    else
        id = userid; /* they picked this one */

    if (id == None)
        return fail("Unable to find window with the requested ID");

    show_properties(d, id, argc - i, &argv[i]);
    
    XCloseDisplay(d);

    return 0;
}
